server/backup_runner.go

```diff
@@ -126,47 +110,85 @@ func (r *BackupRunner) purgeExpiredBackups(ctx context.Context) {
 	}
 
 	for _, instance := range instanceList {
-		if instance.Engine != db.MySQL {
-			continue
-		}
-		backupSettingList, err := r.server.store.FindBackupSetting(ctx, api.BackupSettingFind{InstanceID: &instance.ID})
+		maxRetentionPeriodTs, err := r.getMaxRetentionPeriodTsForMySQLInstance(ctx, instance)
 		if err != nil {
-			log.Error("Failed to find backup settings for instance.", zap.String("instance", instance.Name), zap.Error(err))
-			continue // next instance
+			continue
 		}
-		maxRetentionPeriodForInstance := math.MaxInt
-		for _, bs := range backupSettingList {
-			if bs.RetentionPeriodTs != api.BackupRetentionPeriodUnset && bs.RetentionPeriodTs < maxRetentionPeriodForInstance {
-				maxRetentionPeriodForInstance = bs.RetentionPeriodTs
-			}
+		if maxRetentionPeriodTs == math.MaxInt {
+			log.Info("All the databases in the MySQL instance have unset retention period. Skip deleting binlog files.", zap.String("instance", instance.Name))
+			continue
 		}
-		if maxRetentionPeriodForInstance == math.MaxInt {
-			log.Debug("All the databases in instance have unset retention period. Skip deleting binlog files.", zap.String("instance", instance.Name))
-			continue // next instance
+		log.Debug("Deleting old binlog files for MySQL instance.", zap.String("instance", instance.Name))
+		r.purgeBinlogFiles(ctx, instance.ID, maxRetentionPeriodTs)
+	}
+}
+
+func (r *BackupRunner) getMaxRetentionPeriodTsForMySQLInstance(ctx context.Context, instance *api.Instance) (int, error) {
+	if instance.Engine != db.MySQL {
```

Comment:
Should this be part of instance iterator line 113?

Reply:
SG.